% $Log: abstract.tex,v $
% Revision 1.1  93/05/14  14:56:25  starflt
% Initial revision
%
% Revision 1.1  90/05/04  10:41:01  lwvanels
% Initial revision
%
%
%% The text of your abstract and nothing else (other than comments) goes here.
%% It will be single-spaced and the rest of the text that is supposed to go on
%% the abstract page will be generated by the abstractpage environment.  This
%% file should be \input (not \include 'd) from cover.tex.
In this thesis, I discuss the application and development of methods
for the automated discovery of motifs in sequential data.  These data
include DNA sequences, protein sequences, and real--valued
sequential data such as protein structures and timeseries of
arbitrary dimension. As more genomes are sequenced and annotated,
the need for automated, computational methods for analyzing
biological data is increasing rapidly.  In broad terms, the goal of
this thesis is to treat sequential data sets as unknown languages
and to develop tools for interpreting an understanding these
languages.

The first chapter of this thesis is an introduction to the
fundamentals of motif discovery, which establishes a common mode of
thought and vocabulary for the subsequent chapters.  One of the
central themes of this work is the use of grammatical models,
which are more commonly associated with the field of computational
linguistics.  In the second chapter, I use grammatical models to
design novel antimicrobial peptides (AmPs). AmPs are small proteins
used by the innate immune system to combat bacterial infection in
multicellular eukaryotes. There is mounting evidence that these
peptides are less susceptible to bacterial resistance than
traditional antibiotics and may form the basis for a novel class of
therapeutics. In this thesis, I described the rational design of
novel AmPs that show limited homology to naturally--occurring
proteins but have strong bacteriostatic activity against several
species of bacteria, including \emph{Staphylococcus aureus} and
\emph{Bacillus anthracis}. These peptides were designed using a
linguistic model of natural AmPs by treating the amino acid
sequences of natural AmPs as a formal language and building a set of
regular grammars to describe this language. This set of grammars was
used to create novel, unnatural AmP sequences that conform to the
formal syntax of natural antimicrobial peptides but populate a
previously unexplored region of protein sequence space.

The third chapter describes a novel, GEneric MOtif DIscovery
Algorithm (Gemoda) for sequential data. Gemoda can be applied to any
dataset with a sequential character, including both categorical and
real--valued data.  As I show, Gemoda deterministically discovers
motifs that are maximal in composition and length.  As well, the
algorithm allows any choice of similarity metric for finding motifs.
These motifs are representation--agnostic: they can be represented
using regular expressions, position weight matrices, or 
any other model for sequential data. I demonstrate a
number of applications of the algorithm, including the discovery of
motifs in amino acids and DNA sequences, and the discovery of
conserved protein sub--structures.

The final chapter is devoted to a series of smaller projects,
employing tools and methods indirectly related to motif discovery in
sequential data.  I describe the construction of a software tool,
Biogrep that is designed to match large pattern sets against large
biosequence databases in a \emph{parallel} fashion. This makes
biogrep well--suited to annotating sets of sequences using
biologically significant patterns.  In addition, I show that the
BLOSUM series of amino acid substitution matrices, which are
commonly used in motif discovery and sequence alignment problems,
have changed drastically over time.  The fidelity of amino acid
sequence alignment and motif discovery tools depends strongly on the
target frequencies implied by these underlying matrices.  Thus,
these results suggest that further optimization of these matrices is
possible.

The final chapter also contains two projects wherein I apply
statistical motif discovery tools instead of grammatical tools. In
the first of these two, I develop three different physiochemical
representations for a set of roughly 700 HIV--I protease substrates
and use these representations for sequence classification and
annotation.  In the second of these two projects, I develop a simple
statistical method for parsing out the phenotypic contribution of a
single mutation from libraries of functional diversity that contain
a multitude of mutations and varied phenotypes.  I show that this
new method successfully elucidates the effects of single nucleotide
polymorphisms on the strength of a promoter placed upstream of a
reporter gene.

The central theme, present throughout this work, is the development
and application of novel approaches to finding motifs in sequential
data.  The work on the design of AmPs is very applied and relies
heavily on existing literature.  In contrast, the work on Gemoda is
the greatest contribution of this thesis and contains many new
ideas.
